### DISCUSSION TITLE
test文件中使用handshaking报错问题

### DISCUSSION THREAD

**Discussion by LoserXj** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#37] [Q&A]
![报错](https://user-images.githubusercontent.com/103046776/209545127-2039c90b-88cc-4fcb-8059-69471599431b.png)
请问在test文件的handshaking中handshaking_session.run_grader()这一步报错，具体走到图片画红线地方失败，请问会是什么原因呢

**Comment by TANSixu** [comment]
Hi，请问可以提供一下报错信息以及运行环境[系统+python版本]嘛

**Comment by LoserXj** [comment]
学长你好，运行环境是win，python是3.9
test setup failed
@pytest.fixture(scope='module')
    def normal_session():
        blocking_time = 10
        handshaking_session = grader.GradingSession(grader.normal_handler)
        handshaking_session.add_peer(1, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1, ("127.0.0.1", 48001))
        handshaking_session.add_peer(2, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data2.fragment", 1, ("127.0.0.1", 48002))
        handshaking_session.add_peer(3, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1, ("127.0.0.1", 48003))
        handshaking_session.add_peer(4, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1, ("127.0.0.1", 48004))
        handshaking_session.add_peer(5, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1, ("127.0.0.1", 48005))
        handshaking_session.add_peer(6, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1, ("127.0.0.1", 48006))
        handshaking_session.add_peer(7, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1, ("127.0.0.1", 48007))
>       handshaking_session.run_grader()

D:\PYTHON_PROJECT\CS305-Project-Skeleton\test\basic_handshaking_test.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
D:\PYTHON_PROJECT\CS305-Project-Skeleton\test\grader.py:137: in run_grader
    p.start_peer()
D:\PYTHON_PROJECT\CS305-Project-Skeleton\test\grader.py:31: in start_peer
    self.process = subprocess.Popen(cmd.split(" "), stdin=subprocess.PIPE,stdout=subprocess.DEVNULL,text=True, bufsize=1, universal_newlines=True)
D:\Anaconda3\envs\ExercisePytorch\lib\subprocess.py:951: in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Popen: returncode: None args: ['python3', '-u', 'src/peer.py', '-p', 'test/...>
args = 'python3 -u src/peer.py -p test/tmp1/nodes1.map -c test/tmp1/data1.fragment -m 1 -i 1 -t 60'
executable = None, preexec_fn = None, close_fds = False, pass_fds = ()
cwd = None, env = None
startupinfo = <subprocess.STARTUPINFO object at 0x000001FDE1EB1580>
creationflags = 0, shell = False, p2cread = Handle(924), p2cwrite = 13
c2pread = -1, c2pwrite = Handle(928), errread = -1, errwrite = Handle(932)
unused_restore_signals = True, unused_gid = None, unused_gids = None
unused_uid = None, unused_umask = -1, unused_start_new_session = False

    def _execute_child(self, args, executable, preexec_fn, close_fds,
                       pass_fds, cwd, env,
                       startupinfo, creationflags, shell,
                       p2cread, p2cwrite,
                       c2pread, c2pwrite,
                       errread, errwrite,
                       unused_restore_signals,
                       unused_gid, unused_gids, unused_uid,
                       unused_umask,
                       unused_start_new_session):
        """Execute program (MS Windows version)"""
    
        assert not pass_fds, "pass_fds not supported on Windows."
    
        if isinstance(args, str):
            pass
        elif isinstance(args, bytes):
            if shell:
                raise TypeError('bytes args is not allowed on Windows')
            args = list2cmdline([args])
        elif isinstance(args, os.PathLike):
            if shell:
                raise TypeError('path-like args is not allowed when '
                                'shell is true')
            args = list2cmdline([args])
        else:
            args = list2cmdline(args)
    
        if executable is not None:
            executable = os.fsdecode(executable)
    
        # Process startup details
        if startupinfo is None:
            startupinfo = STARTUPINFO()
        else:
            # bpo-34044: Copy STARTUPINFO since it is modified above,
            # so the caller can reuse it multiple times.
            startupinfo = startupinfo.copy()
    
        use_std_handles = -1 not in (p2cread, c2pwrite, errwrite)
        if use_std_handles:
            startupinfo.dwFlags |= _winapi.STARTF_USESTDHANDLES
            startupinfo.hStdInput = p2cread
            startupinfo.hStdOutput = c2pwrite
            startupinfo.hStdError = errwrite
    
        attribute_list = startupinfo.lpAttributeList
        have_handle_list = bool(attribute_list and
                                "handle_list" in attribute_list and
                                attribute_list["handle_list"])
    
        # If we were given an handle_list or need to create one
        if have_handle_list or (use_std_handles and close_fds):
            if attribute_list is None:
                attribute_list = startupinfo.lpAttributeList = {}
            handle_list = attribute_list["handle_list"] = \
                list(attribute_list.get("handle_list", []))
    
            if use_std_handles:
                handle_list += [int(p2cread), int(c2pwrite), int(errwrite)]
    
            handle_list[:] = self._filter_handle_list(handle_list)
    
            if handle_list:
                if not close_fds:
                    warnings.warn("startupinfo.lpAttributeList['handle_list'] "
                                  "overriding close_fds", RuntimeWarning)
    
                # When using the handle_list we always request to inherit
                # handles but the only handles that will be inherited are
                # the ones in the handle_list
                close_fds = False
    
        if shell:
            startupinfo.dwFlags |= _winapi.STARTF_USESHOWWINDOW
            startupinfo.wShowWindow = _winapi.SW_HIDE
            comspec = os.environ.get("COMSPEC", "cmd.exe")
            args = '{} /c "{}"'.format (comspec, args)
    
        if cwd is not None:
            cwd = os.fsdecode(cwd)
    
        sys.audit("subprocess.Popen", executable, args, cwd, env)
    
        # Start the process
        try:
>           hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                                     # no special security
                                     None, None,
                                     int(not close_fds),
                                     creationflags,
                                     env,
                                     cwd,
                                     startupinfo)
E                                    FileNotFoundError: [WinError 2] 系统找不到指定的文件。

D:\Anaconda3\envs\ExercisePytorch\lib\subprocess.py:1420: FileNotFoundError


============================== 1 error in 0.22s ===============================

进程已结束,退出代码1

**Reply by TANSixu** [reply]
同学好，请留意文档中的配置信息噢：
> This project runs on Ubuntu 20.04, python 3.8. Please be kindly advised that python 3.9 may
introduce error in our testing scripts, and Windows is not supported (theoretically Macs should work,
but we did not try). You can set up your own VM or container locally following varieties of tutorials.
We will also provide remote containers(cpu=2, mem=2G) with already-set development environment
to you if you prefer not to set up your local environment.

**Comment by LoserXj** [comment]
好的好的，不好意思学长，之前以为使用命令行测试全部时必须使用3.8，给的单个test以为什么都可以，我再去试试

**Reply by TANSixu** [reply]
好滴好滴~ 以及可能需要配置一下docker或者虚拟机噢 win上跑会出问题（奇怪的winsocket）

**Comment by venkyqz** [comment]
学长你好，顺便问一下, **我和这位同学遇到的问题类似**, 也是无法正常运行pytest。但是我配置的项目解释器是**Python 3.8.10 WSL (Ubuntu 20.04)**，不知道是不是环境配置的问题，麻烦帮我看看。

以下我的Pycharm启动项设置：

<img width="801" alt="pycharm启动项设置" src="https://user-images.githubusercontent.com/87608363/209906670-735e88b9-d2b2-411f-b2d6-9758b4fc0788.png">

以下是我的报错信息：
test_flooding_whohas :

```
ERROR               [ 25%]
test setup failed
@pytest.fixture(scope='module')
    def normal_session():
        blocking_time = 10
        handshaking_session = grader.GradingSession(grader.normal_handler)
        handshaking_session.add_peer(1, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48001))
        handshaking_session.add_peer(2, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data2.fragment", 1,
                                     ("127.0.0.1", 48002))
        handshaking_session.add_peer(3, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48003))
        handshaking_session.add_peer(4, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48004))
        handshaking_session.add_peer(5, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48005))
        handshaking_session.add_peer(6, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48006))
        handshaking_session.add_peer(7, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48007))
        handshaking_session.run_grader()
    
        handshaking_session.peer_list[("127.0.0.1", 48001)].send_cmd(
            '''DOWNLOAD test/tmp1/download_target.chunkhash test/tmp1/download_result.fragment\n''')
        time.sleep(blocking_time)
    
        for p in handshaking_session.peer_list.values():
>           p.terminate_peer()

test/basic_handshaking_test.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <grader.PeerProc object at 0x7f80b93d19d0>

    def terminate_peer(self):
>       self.process.send_signal(signal.SIGINT)
E       AttributeError: 'function' object has no attribute 'SIGINT'

test/grader.py:56: AttributeError
```

test_send_ihave: 
```
ERROR                    [ 50%]
test setup failed
@pytest.fixture(scope='module')
    def normal_session():
        blocking_time = 10
        handshaking_session = grader.GradingSession(grader.normal_handler)
        handshaking_session.add_peer(1, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48001))
        handshaking_session.add_peer(2, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data2.fragment", 1,
                                     ("127.0.0.1", 48002))
        handshaking_session.add_peer(3, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48003))
        handshaking_session.add_peer(4, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48004))
        handshaking_session.add_peer(5, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48005))
        handshaking_session.add_peer(6, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48006))
        handshaking_session.add_peer(7, "src/peer.py", "test/tmp1/nodes1.map", "test/tmp1/data1.fragment", 1,
                                     ("127.0.0.1", 48007))
        handshaking_session.run_grader()
    
        handshaking_session.peer_list[("127.0.0.1", 48001)].send_cmd(
            '''DOWNLOAD test/tmp1/download_target.chunkhash test/tmp1/download_result.fragment\n''')
        time.sleep(blocking_time)
    
        for p in handshaking_session.peer_list.values():
>           p.terminate_peer()

test/basic_handshaking_test.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <grader.PeerProc object at 0x7f80b93d19d0>

    def terminate_peer(self):
>       self.process.send_signal(signal.SIGINT)
E       AttributeError: 'function' object has no attribute 'SIGINT'

test/grader.py:56: AttributeError
```

**Reply by TANSixu** [reply]
同学你好，请加一下我的qq：846030067，这应该是pycharm的配置问题，咱们一起尝试解决一下。
之前也有一位同学遇到了关于pycharm的问题，由于pycharm的远程逻辑和vscode有些不同，我们正在考虑录制一个视频来帮助大家配置pycharm。

**Comment by novada0** [comment]
学长，能帮忙看看这是什么问题嘛？
![image](https://user-images.githubusercontent.com/107106546/210163788-65819957-f188-4daf-907a-1d71f819c364.png)

**Reply by TANSixu** [reply]
hi，也是默认使用了python2。
请使用这个命令试试：
python3 -m pytest xxxxxxxxxxxx
